### PR TITLE
Reject proposals when they cannot pass

### PR DESCRIPTION
**Pull request by Callum-A**
Addresses Issue #169 

Adds an is_rejected method to both Proposal types which does the same as is_passed however checks no votes for a majority.
Added unit tests  for the logic on the Proposal side.
Added integration tests for both a DAO and Multisig.

Feedback appreciated.

### CODE REVIEW

File: contracts/cw3-multisig/src/state.rs [hunk 1]
```diff
@@ -107,17 +107,17 @@ impl Proposal {
         self.status = self.current_status(block);
     }
 
-    // returns true iff this proposal is sure to pass (even before expiration if no future
-    // sequence of possible votes can cause it to fail)
-    pub fn is_passed(&self, block: &BlockInfo) -> bool {
+    /// Helper function to check if a certain vote count has reached threshold.
+    /// Only called from is_rejected and is_passed for no and yes votes
+    fn does_vote_count_reach_threshold(&self, vote_count: u64, block: &BlockInfo) -> bool {
```
Comment (maurolacy) [inline contracts/cw3-multisig/src/state.rs:112]: Good generalisation. A little unclear what this does, but I think with the comment above, it's OK.

**Review by maurolacy — COMMENTED** [review]
Looks good. ~~It would be good to have unit tests for checking and documenting the "reject early" case.~~ (already done)

Will backport this to cw-plus at some point.

**Review by maurolacy — APPROVED** [review]
Here are the unit tests. It seems I was reviewing an outdated version before.

**Comment by maurolacy** [comment]
Returning to the upstream version of `cw3-flex-multisig` could be a good idea at some point.

Need to confirm / check this, but in principle you could be contributing / adding these there directly there.

**Comment by Callum-A** [comment]
> Returning to the upstream version of `cw3-flex-multisig` could be a good idea at some point.
> 
> Need to confirm / check this, but at least in principle you could be contributing / adding these there directly there.

I can definitely look into this upstream

File: contracts/cw3-multisig/src/state.rs (hunk 1 above)
Comment (Callum-A) [reply contracts/cw3-multisig/src/state.rs:112]: I'm going to look into making it clearer via making the comments better today.

Comment (Callum-A) [reply contracts/cw3-multisig/src/state.rs:119]: Have attempted to improve the comments in my latest commits, feedback welcome.

**Review by ezekiiel — APPROVED** [review]
Looks great. Thanks for all the tests!

**Review by JakeHartnell — APPROVED** [review]

**Comment by JakeHartnell** [comment]
Awesome work!

**Pull request merged**